src/nf/nf_batchnorm_layer_submodule.f90

```diff
@@ -0,0 +1,105 @@
+submodule(nf_batchnorm_layer) nf_batchnorm_layer_submodule
+
+  implicit none
+
+contains
+
+  pure module function batchnorm_layer_cons(num_features) result(res)
+    implicit none
+    integer, intent(in) :: num_features
+    type(batchnorm_layer) :: res
+
+    res % num_features = num_features
+    allocate(res % gamma(num_features), source=1.0)
+    allocate(res % beta(num_features))
+    allocate(res % running_mean(num_features), source=0.0)
+    allocate(res % running_var(num_features), source=1.0)
+    allocate(res % input(num_features, num_features))
+    allocate(res % output(num_features, num_features))
+    allocate(res % gamma_grad(num_features))
+    allocate(res % beta_grad(num_features))
+    allocate(res % input_grad(num_features, num_features))
+
+  end function batchnorm_layer_cons
+
+  module subroutine init(self, input_shape)
+    implicit none
+    class(batchnorm_layer), intent(in out) :: self
+    integer, intent(in) :: input_shape(:)
+
+    self % input = 0
+    self % output = 0
+
+    ! Initialize gamma, beta, running_mean, and running_var
+    self % gamma = 1.0
+    self % beta = 0.0
+    self % running_mean = 0.0
+    self % running_var = 1.0
+
+  end subroutine init
+
+  pure module subroutine forward(self, input)
+    implicit none
+    class(batchnorm_layer), intent(in out) :: self
+    real, intent(in) :: input(:,:)
+
+    ! Store input for backward pass
+    self % input = input
+
+    associate( &
+      ! Normalize the input
+      normalized_input => (input - reshape(self % running_mean, shape(input, 1))) &
+        / sqrt(reshape(self % running_var, shape(input, 1)) + self % epsilon) &
```

Comment:
`running_mean` and `running_var` are not yet updated anywhere, only initialized.